validation.go

```diff
@@ -899,15 +896,32 @@ func (v *subSchema) validateNumber(currentSubSchema *subSchema, value interface{
 
 	// format
 	if currentSubSchema.format != "" {
-		if !FormatCheckers.IsFormat(currentSubSchema.format, float64Value) {
-			result.addInternalError(
-				new(DoesNotMatchFormatError),
-				context,
-				value,
-				ErrorDetails{"format": currentSubSchema.format},
-			)
+		if topSchema.HasFormatCheck(currentSubSchema.format) {
+			applyFmtCheck(topSchema.formatCheckers, currentSubSchema, float64Value, result, context)
+		} else {
+			globalFmtCheck(currentSubSchema, float64Value, result, context)
 		}
 	}
 
 	result.incrementScore()
 }
+
+func globalFmtCheck(currentSubSchema *subSchema, value interface{}, result *Result, context *JsonContext) {
```

Comment:
Ah, that's just my coding style to write these descriptive `subfunctions`. I don't think it's immediately apparent what's happening when it looks something like this:

```go
		if topSchema.HasFormatCheck(currentSubSchema.format) {
			applyFmtCheck(topSchema.formatCheckers, currentSubSchema, float64Value, result, context)
		} else {
			applyFmtCheck(&FormatCheckers, currentSubSchema, float64Value, result, context)
		}
```

Also, I definitely want `applyFmtCheck` subfunction because otherwise it would be repeated x4. 2x in `validateNumber` and then another 2x in `validateString`.

Perhaps I could create  intermediate `structs` to shorten the arg list:

```go
type fmtCheckContext struct {
	value interface{}
	result *Result
	context *JsonContext
}
```

Reply:
I still think you should move the `format` check to `validateCommon`. The only argument that is actually different when you use the `applyFmtCheck` is the first one. You can easily handle that with an if statement. Something along the line of:

```go
	if currentSubSchema.format != "" {
		currentChecker := &FormatCheckers 
		if topSchema.HasFormatCheck(currentSubSchema.format) {
			currentChecker = currentSubSchema.formatCheckers
		}
		if isRightFmt, err := currentChecker.IsFormatWithError(currentSubSchema.format, value); !isRightFmt {
			existingDetails := err.Details()
			if existingDetails == nil {
				existingDetails = ErrorDetails{}
			}
			existingDetails["format"] = currentSubSchema.format
			result.addInternalError(
				err,
				context,
				value,
				existingDetails,
			)
		}		
	}
```

That saves you a lot of code.